File: src-tauri/src/watcher/parsers.rs
use std::path::{Path, PathBuf};

use notify::event::{EventKind, ModifyKind};

use super::types::FileChangeEvent;

/// Map a notify v7 `EventKind` to our change-type string, reproducing Go
/// `MapEventKind` (internal/watcher/parsers.go)'s four outcomes exactly:
///
/// | Go (rjeczalik/notify) | outcome  | notify v7 arm                    |
/// |-----------------------|----------|----------------------------------|
/// | Create                | "add"    | `Create(_)`                      |
/// | Write                 | "change" | `Modify(Data(_))` / `Modify(Any)`|
/// | Remove                | "unlink" | `Remove(_)`                      |
/// | Rename                | "unlink" | `Modify(Name(_))`                |
/// | (default)             | None     | everything else                  |
///
/// notify v7 models a content write as `Modify(Data)` (and, when FSEvents
/// coalesces, `Modify(Any)`), and a rename as `Modify(Name)` — so the Name arm
/// is checked BEFORE the write arms. `Modify(Metadata)` (chmod) is intentionally
/// dropped: Go's portable Event set has no chmod flag, so it maps to None.
pub(crate) fn map_event_kind(kind: &EventKind) -> Option<&'static str> {
    match kind {
        EventKind::Create(_) => Some("add"),
        EventKind::Modify(ModifyKind::Name(_)) => Some("unlink"),
        EventKind::Modify(ModifyKind::Data(_)) | EventKind::Modify(ModifyKind::Any) => {
            Some("change")
        }
        EventKind::Remove(_) => Some("unlink"),
        _ => None,
    }
}

/// Parse a projects-directory file path into a FileChangeEvent, reconciled with
/// Go `ParseProjectFile`.
///
/// Structures (relative to projects_dir):
///   projectId/sessionId.jsonl                        → session file
///   projectId/sessionId/subagents/agent-hash.jsonl   → subagent file
pub(crate) fn parse_project_file(
    projects_dir: &Path,
    file_path: &Path,
    change_type: &str,
) -> Option<FileChangeEvent> {
    // strip_prefix is component-based: a non-descendant (Go's rel starting with
    // "..") returns Err, and file_path == projects_dir (Go's rel == ".") yields
    // an empty remainder → no last component → None below.
    let relative = file_path.strip_prefix(projects_dir).ok()?;
    let components: Vec<&str> = relative
        .components()
        .filter_map(|c| c.as_os_str().to_str())
        .collect();

    let filename = components.last()?;
    if !filename.ends_with(".jsonl") {
        return None;
    }

    let project_id = components.first()?.to_string();

    let (session_id, is_subagent) = match components.len() {
        2 => (filename.strip_suffix(".jsonl")?.to_string(), false),
        4 if components[2] == "subagents" => (components[1].to_string(), true),
        _ => return None,
    };

    Some(FileChangeEvent {
        change_type: change_type.to_string(),
        path: file_path.to_string_lossy().to_string(),
        project_id: Some(project_id),
        session_id: Some(session_id),
        is_subagent,
    })
}

/// Parse a todos-directory file path into a FileChangeEvent, reconciled with Go
/// `ParseTodoFile`. Structure: sessionId.json (flat, non-recursive). No projectId.
pub(crate) fn parse_todo_file(
    todos_dir: &Path,
    file_path: &Path,
    change_type: &str,
) -> Option<FileChangeEvent> {
    let relative = file_path.strip_prefix(todos_dir).ok()?;
    let filename = relative.to_str()?;

    if !filename.ends_with(".json") {
        return None;
    }

    let session_id = filename.strip_suffix(".json")?.to_string();

    Some(FileChangeEvent {
        change_type: change_type.to_string(),
        path: file_path.to_string_lossy().to_string(),
        project_id: None,
        session_id: Some(session_id),
        is_subagent: false,
    })
}

/// Resolve the .claude root, mirroring Go `ResolveClaudeDir`: honour `CLAUDE_ROOT`
/// when it points at an existing path, else fall back to `$HOME/.claude`.
#[cfg_attr(not(test), allow(dead_code))]
pub fn resolve_claude_dir() -> Option<PathBuf> {
    if let Ok(root) = std::env::var("CLAUDE_ROOT") {
        let p = PathBuf::from(root);
        if p.exists() {
            return Some(p);
        }
    }
    dirs::home_dir().map(|h| h.join(".claude"))
}

File: internal/watcher/parsers.go
package watcher

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/rjeczalik/notify"
)

// MapEventKind maps a notify.Event bitmask to a change-type string.
// Returns ("", false) for events we do not care about (e.g. access, chmod).
// Rename is treated as unlink (the old name disappears), matching Rust.
func MapEventKind(e notify.Event) (string, bool) {
	switch {
	case e&notify.Create != 0:
		return "add", true
	case e&notify.Write != 0:
		return "change", true
	case e&notify.Remove != 0:
		return "unlink", true
	case e&notify.Rename != 0:
		return "unlink", true
	default:
		return "", false
	}
}

// ParseProjectFile parses a projects-directory absolute path into a
// FileChangeEvent. Returns nil for paths that don't match known structures.
//
// Supported structures (relative to projectsDir):
//
//	projectId/sessionId.jsonl                        → session file
//	projectId/sessionId/subagents/agent-hash.jsonl   → subagent file
func ParseProjectFile(projectsDir, filePath, changeType string) *FileChangeEvent {
	rel, err := filepath.Rel(projectsDir, filePath)
	if err != nil || strings.HasPrefix(rel, "..") || rel == "." {
		return nil
	}
	parts := strings.Split(filepath.ToSlash(rel), "/")

	filename := parts[len(parts)-1]
	if !strings.HasSuffix(filename, ".jsonl") {
		return nil
	}

	projectID := parts[0]

	var sessionID string
	var isSubagent bool

	switch len(parts) {
	case 2:
		// projectId/sessionId.jsonl
		sessionID = strings.TrimSuffix(filename, ".jsonl")
	case 4:
		// projectId/sessionId/subagents/agent-hash.jsonl
		if parts[2] != "subagents" {
			return nil
		}
		sessionID = parts[1]
		isSubagent = true
	default:
		return nil
	}

	pid := projectID
	sid := sessionID
	return &FileChangeEvent{
		Type:       changeType,
		Path:       filePath,
		ProjectID:  &pid,
		SessionID:  &sid,
		IsSubagent: isSubagent,
	}
}

// ParseTodoFile parses a todos-directory absolute path into a FileChangeEvent.
// Returns nil if the file is not a .json file.
//
// Expected structure: sessionId.json (flat directory, non-recursive watch).
func ParseTodoFile(todosDir, filePath, changeType string) *FileChangeEvent {
	rel, err := filepath.Rel(todosDir, filePath)
	if err != nil || strings.HasPrefix(rel, "..") || rel == "." {
		return nil
	}
	// filepath.ToSlash ensures "/" even on Windows (future-proofing).
	filename := filepath.ToSlash(rel)
	if !strings.HasSuffix(filename, ".json") {
		return nil
	}
	sessionID := strings.TrimSuffix(filename, ".json")
	sid := sessionID
	return &FileChangeEvent{
		Type:      changeType,
		Path:      filePath,
		SessionID: &sid,
	}
}

// ResolveClaudeDir returns the .claude root directory, checking CLAUDE_ROOT
// env var first. Mirrors lifecycle.rs::resolve_claude_dir.
func ResolveClaudeDir() (string, bool) {
	if root := os.Getenv("CLAUDE_ROOT"); root != "" {
		if _, err := os.Stat(root); err == nil {
			return root, true
		}
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", false
	}
	return filepath.Join(home, ".claude"), true
}
